File: src/rails/game/correct/CashCorrectionManager.java
package rails.game.correct;

import java.util.List;

import rails.common.DisplayBuffer;
import rails.common.LocalText;
import rails.common.ReportBuffer;
import rails.game.GameManager;
import rails.game.Currency;
import rails.game.MoneyOwner;
import rails.game.Player;
import rails.game.PublicCompany;
import rails.game.state.ChangeStack;

public class CashCorrectionManager extends CorrectionManager {
    
    private CashCorrectionManager(GameManager parent) {
        super(parent, CorrectionType.CORRECT_CASH);
    }
    
    public static CashCorrectionManager create(GameManager parent) {
        return new CashCorrectionManager(parent);
    }
    
    @Override
    public List<CorrectionAction> createCorrections() {
        List<CorrectionAction> actions = super.createCorrections();
        
        if (isActive()) {
            List<Player> players = getRoot().getPlayerManager().getPlayers();
            for(Player pl:players){
                actions.add(new CashCorrectionAction(pl));
            }

            List<PublicCompany> publicCompanies = getParent().getAllPublicCompanies();
            for(PublicCompany pc:publicCompanies){
                if (pc.hasFloated() && !pc.isClosed())
                    actions.add(new CashCorrectionAction(pc));
            }
        }

        return actions;
    }
    
    @Override
    public boolean executeCorrection(CorrectionAction action){
        if (action instanceof CashCorrectionAction)
            return execute((CashCorrectionAction) action);
        else
             return super.executeCorrection(action);
    }

    private boolean execute(CashCorrectionAction cashAction) {

        boolean result = false;

        MoneyOwner ch = cashAction.getCashHolder();
        int amount = cashAction.getAmount();

        String errMsg = null;

        while (true) {
            if (amount == 0 ) {
                errMsg =
                    LocalText.getText("CorrectCashZero");
                break;
            }
            if ((amount + ch.getCash()) < 0) {
                errMsg =
                    LocalText.getText("NotEnoughMoney", 
                            ch.getId(),
                            Currency.format(this, ch.getCash()),
                            Currency.format(this, -amount) 
                    );
                break;
            }
            break;   
        }

        if (errMsg != null) {
            DisplayBuffer.add(this, LocalText.getText("CorrectCashError",
                    ch.getId(),
                    errMsg));
            result = true;
        } else {
            // no error occured 
            
            ChangeStack.start(this, cashAction);

            String msg;
            if (amount < 0) {
                // negative amounts: remove cash from cashholder
                String text = Currency.toBank(ch, -amount);

                msg = LocalText.getText("CorrectCashSubstractMoney",
                        ch.getId(),
                        text );
            } else {
                // positive amounts: add cash to cashholder
                String text = Currency.fromBank(amount, ch);
                msg = LocalText.getText("CorrectCashAddMoney",
                        ch.getId(),
                        text);
            }
            ReportBuffer.add(this, msg);
            getParent().addToNextPlayerMessages(msg, true);
            result = true;
        }
    
       return result;
    }

}

File: dotnet/RailsLib.Net/Rails/Game/Correct/CashCorrectionManager.cs
using GameLib.Net.Common;
using GameLib.Net.Game;
using GameLib.Net.Game.Financial;
using GameLib.Net.Game.State;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Text;

namespace GameLib.Rails.Game.Correct
{
    [JsonObject(MemberSerialization.Fields)]
    public class CashCorrectionManager : CorrectionManager
    {
        private CashCorrectionManager(GameManager parent) : base(parent, CorrectionType.CORRECT_CASH)
        {
        }

        public static CashCorrectionManager Create(GameManager parent)
        {
            return new CashCorrectionManager(parent);
        }

        override public List<CorrectionAction> CreateCorrections()
        {
            List<CorrectionAction> actions = base.CreateCorrections();

            if (IsActive())
            {
                IReadOnlyCollection<Player> players = GetRoot.PlayerManager.Players;
                foreach (Player pl in players)
                {
                    actions.Add(new CashCorrectionAction(pl));
                }

                IReadOnlyCollection<PublicCompany> publicCompanies = Parent.GetAllPublicCompanies();
                foreach (PublicCompany pc in publicCompanies)
                {
                    if (pc.HasFloated() && !pc.IsClosed())
                        actions.Add(new CashCorrectionAction(pc));
                }
            }

            return actions;
        }

        override public bool ExecuteCorrection(CorrectionAction action)
        {
            if (action is CashCorrectionAction)
                return Execute((CashCorrectionAction)action);
            else
                return base.ExecuteCorrection(action);
        }

        private bool Execute(CashCorrectionAction cashAction)
        {

            bool result = false;

            IMoneyOwner ch = cashAction.CashHolder;
            int amount = cashAction.Amount;

            string errMsg = null;

            while (true)
            {
                if (amount == 0)
                {
                    errMsg =
                        LocalText.GetText("CorrectCashZero");
                    break;
                }
                if ((amount + ch.Cash) < 0)
                {
                    errMsg =
                        LocalText.GetText("NotEnoughMoney",
                                ch.Id,
                                Bank.Format(this, ch.Cash),
                                Bank.Format(this, -amount));
                    break;
                }
                break;
            }

            if (errMsg != null)
            {
                DisplayBuffer.Add(this, LocalText.GetText("CorrectCashError",
                        ch.Id,
                        errMsg));
                result = true;
            }
            else
            {
                // no error occurred 
                string msg;
                if (amount < 0)
                {
                    // negative amounts: remove cash from cashholder
                    string text = Currency.ToBank(ch, -amount);

                    msg = LocalText.GetText("CorrectCashSubstractMoney",
                            ch.Id,
                            text);
                }
                else
                {
                    // positive amounts: add cash to cashholder
                    string text = Currency.FromBank(amount, ch);
                    msg = LocalText.GetText("CorrectCashAddMoney",
                            ch.Id,
                            text);
                }
                ReportBuffer.Add(this, msg);
                Parent.AddToNextPlayerMessages(msg, true);
                result = true;
            }

            return result;
        }
    }
}
